src/Service/MailTemplateHelper.php

```diff
@@ -0,0 +1,247 @@
+     * @param $entity
+     *
+     * @return array|false|mixed|string[]
+     */
+    private function getValues(?object $entity, TemplateProcessor $templateProcessor)
+    {
+        $placeHolders = $templateProcessor->getVariables();
+
+        $values = [];
+
+        if (null === $entity) {
+            // We don't have any data; highlight placeholders with colored markers.
+            $values = array_combine(
+                $placeHolders,
+                array_map(static function ($name) {
+                    $marker = new TextRun();
+                    $marker->addText('${'.$name.'}', ['bgColor' => 'FFFF99']);
+
+                    return $marker;
+                }, $placeHolders)
+            );
+        } else {
+            // Serialize to csv to flatten array values and get keys separated by `.`, i.e.
+            //   ['name' => ['given' => 'Anders', 'family' => 'And']]
+            // will be converted to
+            //   ['name.given' => 'Anders', 'name.family' => 'And']
+            $csv = $this->serializer->serialize($entity, 'csv', ['groups' => ['mail_template']]);
+            // We now have a csv string with two lines (the first is the header) that we split and parse.
+            [$header, $row] = array_map('str_getcsv', explode(PHP_EOL, $csv, 2));
+            /**
+             * @ psalm-suppress InvalidArgument
+             *
+             * Argument 1 of array_combine expects array<array-key, array-key>, non-empty-list<null|string> provided (see https://psalm.dev/004)
+             */
+            $values = array_combine($header, $row);
+        }
+
+        // Add some default values.
+        $values['now'] = (new \DateTimeImmutable('now'))->format(\DateTimeImmutable::ATOM);
+        $values['today'] = (new \DateTimeImmutable('today'))->format(\DateTimeImmutable::ATOM);
+
+        foreach ($placeHolders as $placeHolder) {
+            if (preg_match('/^(?P<key>[^:]+):(?P<format>.+)$/', $placeHolder, $matches)) {
+                [, $key, $format] = $matches;
+                if (isset($values[$key])) {
+                    $values[$placeHolder] = $this->formatValue($values[$key], $format);
+                }
+            }
+        }
+
+        // Set empty string values for all template variables without a value.
+        $values += array_map(static function ($count) { return ''; }, $templateProcessor->getVariableCount());
+
+        return $values;
+    }
+
+    /**
+     * Format a value.
+     */
+    private function formatValue(string $value, string $format): string
+    {
+        try {
+            $date = new \DateTimeImmutable($value);
+            $formatter = new \IntlDateFormatter(\Locale::getDefault(), \IntlDateFormatter::NONE, \IntlDateFormatter::NONE);
+            // https://unicode-org.github.io/icu/userguide/format_parse/datetime/#datetime-format-syntax
+            $formatter->setPattern($format);
+
+            return $formatter->format($date);
+        } catch (\Exception $exception) {
```

Comment:
Maybe catch specific exceptions?

Reply:
`new \DateTimeImmutable()` throws an `\Exception` on error, so I cannot really be more specific in my catch clause. Clarifying comment added in https://github.com/itk-dev/naevnssekretariatet/pull/61/commits/ad9b9c068958ea6fc7934a2daa7e4dd406cf946d.